fem/ceed_algebraic.cpp

```diff
@@ -0,0 +1,621 @@
+   }
+   else
+   {
+      const double jacobi_scale = 0.65;
+      out = new OperatorJacobiSmoother(t_diag, ess_tdofs, jacobi_scale);
+   }
+   CeedVectorDestroy(&diagceed);
+   return out;
+}
+
+#ifdef MFEM_USE_MPI
+
+class CeedAMG : public Solver
+{
+public:
+   CeedAMG(ConstrainedMFEMCeedOperator &oper, HypreParMatrix *P)
+   {
+      MFEM_ASSERT(P != NULL, "");
+      const Array<int> ess_tdofs = oper.GetEssentialTrueDofs();
+      height = width = oper.Height();
+
+      CeedOperatorFullAssemble(oper.GetCeedOperator(), &mat_local);
+
+      {
+         HypreParMatrix hypre_local(
+            P->GetComm(), P->GetGlobalNumRows(), P->RowPart(), mat_local);
+         op_assembled = RAP(&hypre_local, P);
+      }
+      HypreParMatrix *mat_e = op_assembled->EliminateRowsCols(ess_tdofs);
+      delete mat_e;
+      amg = new HypreBoomerAMG(*op_assembled);
+      amg->SetPrintLevel(0);
+   }
+   void SetOperator(const Operator &op) { }
+   void Mult(const Vector &x, Vector &y) const { amg->Mult(x, y); }
+   ~CeedAMG()
+   {
+      delete op_assembled;
+      delete amg;
+      delete mat_local;
+   }
+private:
+   SparseMatrix *mat_local;
+   HypreParMatrix *op_assembled;
+   HypreBoomerAMG *amg;
+};
+
+#endif
+
+void CoarsenEssentialDofs(const Operator &interp,
+                          const Array<int> &ho_ess_tdofs,
+                          Array<int> &alg_lo_ess_tdofs)
+{
+   Vector ho_boundary_ones(interp.Height());
+   ho_boundary_ones = 0.0;
+   const int *ho_ess_tdofs_h = ho_ess_tdofs.HostRead();
+   for (int i=0; i<ho_ess_tdofs.Size(); ++i)
+   {
+      ho_boundary_ones[ho_ess_tdofs_h[i]] = 1.0;
+   }
+   Vector lo_boundary_ones(interp.Width());
+   interp.MultTranspose(ho_boundary_ones, lo_boundary_ones);
+   auto lobo = lo_boundary_ones.HostRead();
+   for (int i = 0; i < lo_boundary_ones.Size(); ++i)
+   {
+      if (lobo[i] > 0.9)
+      {
+         alg_lo_ess_tdofs.Append(i);
+      }
+   }
+}
+
+template <typename INTEG>
+int TryToAddCeedSubOperator(BilinearFormIntegrator *integ_in, CeedOperator op)
+{
+   INTEG *integ = dynamic_cast<INTEG*>(integ_in);
+   if (integ != NULL)
+   {
+      CeedCompositeOperatorAddSub(op, integ->GetCeedData()->oper);
+      return 1;
+   }
+   return 0;
+}
+
+CeedOperator CreateCeedCompositeOperatorFromBilinearForm(BilinearForm &form)
+{
+   CeedOperator op;
+   CeedCompositeOperatorCreate(internal::ceed, &op);
+
+   // Get the domain bilinear form integrators (DBFIs)
+   Array<BilinearFormIntegrator*> *bffis = form.GetDBFI();
+   int num_integrators = bffis->Size();
+
+   int count = 0;
+   for (int i = 0; i < num_integrators; ++i)
+   {
+      BilinearFormIntegrator *integ = (*bffis)[i];
+      count += TryToAddCeedSubOperator<DiffusionIntegrator>(integ, op);
+      count += TryToAddCeedSubOperator<MassIntegrator>(integ, op);
+      count += TryToAddCeedSubOperator<VectorDiffusionIntegrator>(integ, op);
+      count += TryToAddCeedSubOperator<VectorMassIntegrator>(integ, op);
+   }
```

Comment:
Should all libCEED supported `Integrator` be tested here?

Reply:
Yes, I think so, what am I missing? Convection is not merged yet, that's the only one I know of.

Reply:
It just has me thinking, it's bug prone (or omission prone), and it's silent if it fails, I would think that a virtual method in `NonlinearIntegrator` might make more sense, was this choice made to avoid polluting the `Integrator`s' code?